torch/profiler/utils.py

```diff
@@ -0,0 +1,116 @@
+from collections import deque
+from dataclasses import dataclass
+from torch.profiler import profile, DeviceType
+import torch
+
+
+class EventKey:
+    def __init__(self, event):
+        self.event = event
+
+    def __hash__(self):
+        return hash(self.event.id)
+
+    def __eq__(self, other):
+        return self.event.id == other.event.id
+
+    def __repr__(self):
+        return f"<{self.event.name()} id={self.event.correlation_id}>"
+
+
+@dataclass
+class EventMetrics:
+    duration_time_ns: int = 0
+    self_time_ns: int = 0
+    idle_time_ns: int = 0
+
+    def percent_idle_time(self):
+        if self.duration_time_ns == 0:
+            return 0
+        return self.idle_time_ns / self.duration_time_ns
+
+
+def compute_event_metrics(prof: torch.autograd.profiler.profile):
+    metrics = dict()
+    compute_self_time(prof, metrics)
+    compute_idle_time(prof, metrics)
+    return metrics
+
+
+def compute_self_time(prof, metrics):
+    '''
+   Computes event's self time(total time - time in child ops).
+
+       Parameters:
+            event_tree: Profiler's kineto_results.experimental_event_tree
+    '''
+    stack = deque()
+    event_tree = prof.kineto_results.experimental_event_tree()
+    for event in event_tree:
+        stack.append(event)
+
+    # standard iterating dfs
+    while stack:
+        curr_event = stack.pop()
+        self_time = curr_event.duration_time_ns
+        if curr_event.children:
+            for child_event in curr_event.children:
+                self_time - child_event.duration_time_ns
+                stack.append(child_event)
+        if EventKey(curr_event) in metrics:
+            metrics[EventKey(curr_event)].self_time_us = self_time
+        else:
+            metrics[EventKey(curr_event)] = EventMetrics(self_time_ns=self_time)
+        metrics[EventKey(curr_event)].duration_time_ns = curr_event.duration_time_ns
+
+
+def compute_idle_time(prof: torch.autograd.profiler.profile, metrics):
+    event_tree = prof.kineto_results.experimental_event_tree()
+    event_list = prof.kineto_results.events()
+
+    def is_cuda_launch_kernel(e):
+        return e.name() == "cudaLaunchKernel"
+
+    def is_cuda_kernel(e):
+        # TODO: find a better way to identify cudaLaunchKernel
+        return e.device_type() == DeviceType.CUDA and e.name() != "[memory]" and e.name() != "Memset (Device)"
+
+    # Record All the idle intervals
+    queue_depth = 0
+    idle_interval = []
+    cuda_kernel_events = [event for event in event_list if is_cuda_launch_kernel(event) or is_cuda_kernel(event)]
+    cuda_kernel_events.sort(key=lambda e: e.start_us())
+
+    if len(cuda_kernel_events) == 1:
+        return
+    for i in range(1, len(cuda_kernel_events)):
```

Comment:
Neat little python-ism:
```
for prev_event, curr_event in zip(cuda_kernel_events, cuda_kernel_events[1:]):
```

That will save you from both having to do the `len(cuda_kernel_events) == 1` check and indexing.